File: exercise/day15/src/main/java/document/DocumentTemplateType.java
package document;

import static document.RecordType.ALL;

public enum DocumentTemplateType {
    DEERPP("DEER", RecordType.INDIVIDUAL_PROSPECT),
    DEERPM("DEER", RecordType.LEGAL_PROSPECT),
    AUTP("AUTP", RecordType.INDIVIDUAL_PROSPECT),
    AUTM("AUTM", RecordType.LEGAL_PROSPECT),
    SPEC("SPEC", RecordType.ALL),
    GLPP("GLPP", RecordType.INDIVIDUAL_PROSPECT),
    GLPM("GLPM", RecordType.LEGAL_PROSPECT);

    private final String documentType;
    private final RecordType recordType;

    DocumentTemplateType(String documentType, RecordType recordType) {
        this.documentType = documentType;
        this.recordType = recordType;
    }

    public static DocumentTemplateType fromDocumentTypeAndRecordType(String documentType, String recordType) {
        for (DocumentTemplateType dtt : DocumentTemplateType.values()) {
            if (dtt.getDocumentType().equalsIgnoreCase(documentType)
                    && dtt.getRecordType().equals(RecordType.valueOf(recordType))) {
                return dtt;
            } else if (dtt.getDocumentType().equalsIgnoreCase(documentType)
                    && dtt.getRecordType().equals(ALL)) {
                return dtt;
            }
        }
        throw new IllegalArgumentException("Invalid Document template type or record type");
    }

    private RecordType getRecordType() {
        return recordType;
    }

    private String getDocumentType() {
        return documentType;
    }
}

File: exercise/kotlin/day15/src/main/kotlin/document/DocumentTemplateType.kt
package document

import document.RecordType.*

enum class DocumentTemplateType(val documentType: String, val recordType: RecordType) {
    DEERPP("DEER", INDIVIDUAL_PROSPECT),
    DEERPM("DEER", LEGAL_PROSPECT),
    AUTP("AUTP", INDIVIDUAL_PROSPECT),
    AUTM("AUTM", LEGAL_PROSPECT),
    SPEC("SPEC", ALL),
    GLPP("GLPP", INDIVIDUAL_PROSPECT),
    GLPM("GLPM", LEGAL_PROSPECT);

    companion object {
        fun fromDocumentTypeAndRecordType(documentType: String, recordType: String): DocumentTemplateType {
            for (dtt in entries) {
                if (dtt.documentType.equals(documentType, ignoreCase = true)
                    && dtt.recordType == RecordType.valueOf(recordType)
                ) {
                    return dtt
                } else if (dtt.documentType.equals(
                        documentType,
                        ignoreCase = true
                    ) && dtt.recordType == ALL
                ) {
                    return dtt
                }
            }
            throw IllegalArgumentException("Invalid Document template type or record type")
        }
    }
}